src/Common/PoolWithFailoverBase.h

```diff
@@ -93,6 +93,18 @@ class PoolWithFailoverBase : private boost::noncopyable
         double staleness = 0.0; /// Helps choosing the "least stale" option when all replicas are stale.
     };
 
+    struct PoolState;
+
+    using PoolStates = std::vector<PoolState>;
+
+    struct ShuffledPool
+    {
+        NestedPool * pool{};
+        const PoolState * state{};
+        size_t index = 0;
+        size_t error_count = 0;
+    };
+
```

Comment:
Can't we make it all `protected`?

Reply:
ShuffledPool is used in HedgedConnectionsFactory